// @restart/context/es/mapContextToProps.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

import React from 'react';
import forwardRef from './forwardRef';

const getDisplayName = Component => {
  const name = typeof Component === 'string' ? Component : Component.name || Component.displayName;
  return name ? `ContextTransform(${name})` : 'ContextTransform';
};

const ensureConsumer = c => c.Consumer || c;

function $mapContextToProps(_ref, Component) {
  let maybeArrayOfConsumers = _ref.consumers,
      mapToProps = _ref.mapToProps,
      displayName = _ref.displayName,
      _ref$forwardRefAs = _ref.forwardRefAs,
      forwardRefAs = _ref$forwardRefAs === void 0 ? 'ref' : _ref$forwardRefAs;
  let consumers = maybeArrayOfConsumers;

  if (!Array.isArray(maybeArrayOfConsumers)) {
    consumers = [maybeArrayOfConsumers];
  }

  const SingleConsumer = ensureConsumer(consumers[0]);

  function singleRender(props, ref) {
    const propsWithRef = _extends({
      [forwardRefAs]: ref
    }, props);

    return React.createElement(SingleConsumer, null, value => React.createElement(Component, _extends({}, propsWithRef, mapToProps(value, props))));
  }

  function multiRender(props, ref) {
    const propsWithRef = _extends({
      [forwardRefAs]: ref
    }, props);

    return consumers.reduceRight((inner, Context) => function () {
      for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
        args[_key] = arguments[_key];
      }

      const Consumer = ensureConsumer(Context);
      return React.createElement(Consumer, null, value => inner(...args, value));
    }, function () {
      for (var _len2 = arguments.length, contexts = new Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
        contexts[_key2] = arguments[_key2];
      }

      return React.createElement(Component, _extends({}, propsWithRef, mapToProps(...contexts, props)));
    })();
  }

  const contextTransform = consumers.length === 1 ? singleRender : multiRender;
  return forwardRef(contextTransform, {
    displayName: displayName || getDisplayName(Component)
  });
}

export default function mapContextToProps(maybeOpts, mapToProps, Component) {
  if (arguments.length === 2) return $mapContextToProps(maybeOpts, mapToProps);
  return $mapContextToProps({
    consumers: maybeOpts,
    mapToProps
  }, Component);
}
